curvefs/src/volume/block_device_aio.cpp

```diff
@@ -0,0 +1,292 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * Project: curve
+ * Date: Thursday Jul 07 14:59:19 CST 2022
+ * Author: wuhanqing
+ */
+
+#include "curvefs/src/volume/block_device_aio.h"
+
+#include <atomic>
+#include <cstring>
+#include <mutex>
+
+#include "absl/memory/memory.h"
+#include "include/client/libcurve.h"
+#include "include/client/libcurve_define.h"
+#include "src/client/libcurve_file.h"
+#include "src/common/fast_align.h"
+
+namespace curvefs {
+namespace volume {
+
+using ::curve::common::align_down;
+using ::curve::common::align_up;
+using ::curve::common::is_aligned;
+
+namespace {
+
+const char* ToString(LIBCURVE_OP op) {
+    switch (op) {
+        case LIBCURVE_OP_READ:
+            return "Read";
+        case LIBCURVE_OP_WRITE:
+            return "Write";
+        case LIBCURVE_OP_DISCARD:
+            return "Discard";
+        default:
+            return "Unknown";
+    }
+}
+
+std::ostream& operator<<(std::ostream& os, CurveAioContext* aio) {
+    os << "[off: " << aio->offset << ", len: " << aio->length
+       << ", ret: " << aio->ret << ", type: " << ToString(aio->op) << "]";
+
+    return os;
+}
+
+void AioReadCallBack(CurveAioContext* aio) {
+    AioRead* read = reinterpret_cast<AioRead*>(reinterpret_cast<char*>(aio) -
+                                               offsetof(AioRead, aio));
+
+    std::lock_guard<std::mutex> lock(read->mtx);
+    read->done = true;
+    read->cond.notify_one();
+}
+
+void AioWriteCallBack(CurveAioContext* aio) {
+    AioWrite* write = reinterpret_cast<AioWrite*>(reinterpret_cast<char*>(aio) -
+                                                  offsetof(AioWrite, aio));
+
+    std::lock_guard<std::mutex> lock(write->mtx);
+    write->done = true;
+    write->cond.notify_one();
+}
+
+void AioWritePaddingReadCallBack(CurveAioContext* aio) {
+    AioWrite::PaddingRead* padding = reinterpret_cast<AioWrite::PaddingRead*>(
+        reinterpret_cast<char*>(aio) - offsetof(AioWrite::PaddingRead, aio));
+
+    padding->base->OnPaddingReadComplete(aio);
+}
+
+}  // namespace
+
+AioRead::AioRead(off_t offset,
+                 size_t length,
+                 char* data,
+                 FileClient* dev,
+                 int fd)
+    : aio(), offset(offset), length(length), data(data), dev(dev), fd(fd) {}
+
+void AioRead::Issue() {
+    if (is_aligned(offset, IO_ALIGNED_BLOCK_SIZE) &&
+        is_aligned(length, IO_ALIGNED_BLOCK_SIZE)) {
+        aio.ret = -1;
+        aio.op = LIBCURVE_OP_READ;
+        aio.cb = AioReadCallBack;
+        aio.offset = offset;
+        aio.length = length;
+        aio.buf = data;
+
+        int ret = dev->AioRead(fd, &aio);
+        if (ret < 0) {
+            LOG(ERROR) << "Failed to issue aio read: " << &aio;
+            done = true;
```

Comment:
need atomic ?

Reply:
in this case, it means we failed to start an async read request, and we mark this request as failed, so, there are no other threads that will notify it, only `Wait` will read it and is also called in the same thread.